File: Ch07/SecureNote/SecureNote/DetailViewController.m
//
//  DetailViewController.m
//  SecureNote
//
//  Created by Rob Warner on 10/13/14.
//  Copyright (c) 2014 Michael Privat and Rob Warner. All rights reserved.
//

#import "DetailViewController.h"
#import "Note.h"

@implementation DetailViewController

#pragma mark - Managing the note

- (void)setNote:(Note *)note {
  if (_note != note) {
    _note = note;
    [self configureView];
  }
}

- (void)configureView {
  if (self.note) {
    self.noteTitle.text = self.note.title;
    self.noteBody.text = self.note.body;
  } else {
    self.noteTitle.text = @"";
    self.noteBody.text = @"";
  }
}

- (void)viewDidLoad {
  [super viewDidLoad];
  [self configureView];
}

- (void)viewWillDisappear:(BOOL)animated {
  self.note.title = self.noteTitle.text;
  self.note.body = self.noteBody.text;
  
  NSError *error;
  if (![[self.note managedObjectContext] save:&error]) {
    NSLog(@"Error saving note %@ -- %@ %@", self.noteTitle.text, error, [error userInfo]);
  }
}

@end

File: Ch07/SecureNoteSwift/SecureNoteSwift/DetailViewController.swift
//
//  DetailViewController.swift
//  SecureNoteSwift
//
//  Created by Michael Privat on 10/8/14.
//  Copyright (c) 2014 Pro iOS Persistence. All rights reserved.
//

import UIKit

class DetailViewController: UIViewController {
    @IBOutlet weak var noteTitle: UITextField!
    @IBOutlet weak var noteBody: UITextView!
    
    var note: Note? {
        didSet {
            self.configureView()
        }
    }

    func configureView() {
        if self.noteTitle != nil {
            if let note = self.note {
                self.noteTitle.text = note.title
                self.noteBody.text = note.body
            }
            else {
                self.noteTitle.text = ""
                self.noteBody.text = ""
            }
        }
    }

    override func viewDidLoad() {
        super.viewDidLoad()
        self.configureView()
    }

    override func viewWillDisappear(animated: Bool) {
        self.note?.title = self.noteTitle.text
        self.note?.body = self.noteBody.text
      
        if let managedObjectContext = self.note?.managedObjectContext {
            var error: NSError? = nil
            if !managedObjectContext.save(&error) {
              NSLog("Error saving note \(self.noteTitle.text) -- \(error) \(error!.userInfo)")
            }
        }
    }
    
    func setNote(note: Note) {
        if(self.note != note) {
            self.note = note
            self.configureView()
        }
    }
}
